$namespaces:
  sbg: https://sevenbridges.com
arguments:
- position: 0
  prefix: ''
  shellQuote: false
  valueFrom: "${\n    var in_alignments = [].concat(inputs.in_alignments)[0];\n  \
    \  var output_ext    = inputs.compress_outputs ? \".fastq.gz\" : \".fastq\";\n\
    \    var interleave    = inputs.interleave;\n    var output_prefix = ''; \n  \
    \  var cmd_line      = '';\n    cmd_line          = \"cmd='' && paired_end=`samtools\
    \ view -h \" + in_alignments.path + \" | head -n 500000 | samtools view -Sc -f\
    \ 0x1 -`\";\n\n  if (!inputs.outputs_by_readgroup)\n    {\n        if (inputs.output_prefix)\n\
    \        {\n            output_prefix = inputs.output_prefix;\n        }\n   \
    \     else\n        {\n            if (in_alignments.metadata && in_alignments.metadata.sample_id)\n\
    \            {\n                output_prefix = in_alignments.metadata.sample_id;\n\
    \            }\n            else\n            {\n                output_prefix\
    \ = in_alignments.path.split('/').pop().split('.')[0];\n            }        \
    \  \n        }           \n        \n        cmd_line = cmd_line + \" && if [\
    \ $paired_end != 0 ]; then cmd='--FASTQ \" + output_prefix; \n        \n     \
    \   if (interleave)\n        {\n            cmd_line = cmd_line + \".interleaved\"\
    \ + output_ext + \"';\";\n        }\n        else\n        {\n            cmd_line\
    \ = cmd_line + \".pe_1\" + output_ext;\n            cmd_line = cmd_line + \" --SECOND_END_FASTQ\
    \ \" + output_prefix + \".pe_2\" + output_ext;\n            cmd_line = cmd_line\
    \ + \" --UNPAIRED_FASTQ \" + output_prefix + \".unpaired\" + output_ext + \"';\"\
    ;\n        }        \n        cmd_line = cmd_line + \" else cmd='--FASTQ \" +\
    \ output_prefix  + \".se\" + output_ext + \"'; fi;\";\n        return cmd_line;\n\
    \    }\n    else\n    {\n        return \"cmd='--OUTPUT_DIR .'\";\n    }\n}\n\n"
- position: 1
  prefix: ''
  shellQuote: false
  valueFrom: /opt/gatk
- position: 2
  prefix: ''
  shellQuote: false
  valueFrom: "${\n    if (inputs.memory_per_job)\n    {\n        return \"--java-options\"\
    ;\n    }\n    else {\n        return '';\n    }\n    \n}"
- position: 3
  prefix: ''
  shellQuote: false
  valueFrom: "${\n    if (inputs.memory_per_job) {\n        return '\\\"-Xmx'.concat(inputs.memory_per_job,\
    \ 'M') + '\\\"';\n    }\n    else {\n        return \"\";\n    }\n}"
- position: 4
  prefix: ''
  shellQuote: false
  valueFrom: SamToFastq
- position: 6
  prefix: ''
  shellQuote: false
  valueFrom: "${\n        return '$cmd';\n}"
baseCommand: []
class: CommandLineTool
cwlVersion: v1.0
doc: "The **GATK SamToFastq** tool converts a SAM or BAM file to FASTQ.\n\nThis tool\
  \ extracts read sequences and qualities from the input SAM/BAM file and writes them\
  \ into the output file in Sanger FASTQ format.\n\nIn the RC mode (default is True),\
  \ if the read is aligned and the alignment is to the reverse strand on the genome,\
  \ the read sequence from input SAM file will be reverse-complemented prior to writing\
  \ it to FASTQ in order to correctly restore the original read sequence as it was\
  \ generated by the sequencer [1].\n\n*A list of **all inputs and parameters** with\
  \ corresponding descriptions can be found at the bottom of the page.*\n\n###Common\
  \ Use Cases\n\n* The **GATK SamToFastq** tool requires a BAM/SAM file on its **Input\
  \ BAM/SAM file** (`--INPUT`) input. The tool generates a single-end FASTQ file on\
  \ its **Output FASTQ file(s)** output if the input BAM/SAM file is single end. In\
  \ case the input file is paired end, the tool outputs the first end of the pair\
  \ FASTQ and the second end of the pair FASTQ on its **Output FASTQ file(s)** output,\
  \ except when the **Interleave** (`--INTERLEAVE`) option is set to True. If the\
  \ output is an interleaved FASTQ file, if paired, each line will have /1 or /2 to\
  \ describe which end it came from.\n\n* The **GATK SamToFastq** tool supports an\
  \ optional parameter  **Output by readgroup** (`--OUTPUT_BY_READGROUP`) which, when\
  \ true, outputs a FASTQ file per read group (two FASTQ files per read group if the\
  \ group is paired).\n\n* Usage example (input BAM file is single-end):\n\n```\n\
  gatk SamToFastq \n     --INPUT input.bam\n     --FASTQ output.fastq\n```\n\n\n\n\
  \n\n* Usage example (input BAM file is paired-end):\n\n```\ngatk SamToFastq \n \
  \    --INPUT input.bam\n     --FASTQ output.pe_1.fastq\n     --SECOND_END_FASTQ\
  \ output.pe_2.fastq\n     --UNPAIRED_FASTQ unpaired.fastq\n\n```\n\n###Changes Introduced\
  \ by Seven Bridges\n\n* The GATK SamToFastq tool is implemented to check if the\
  \ input alignments file contains single-end or paired-end data and according to\
  \ that generates different command lines for these two modes and thus produces appropriate\
  \ output files on its **Output FASTQ file(s)** output (one FASTQ file in single-end\
  \ mode and two FASTQ files if the input alignment file contains paired-end data).\
  \ \n\n* All output files will be prefixed using the **Output prefix** parameter.\
  \ In case the **Output prefix** is not provided, the output prefix will be the same\
  \ as the Sample ID metadata from the **input SAM/BAM file**, if the Sample ID metadata\
  \ exists. Otherwise, the output prefix will be inferred from the **Input SAM/BAM**\
  \ filename. This way, having identical names of the output files between runs is\
  \ avoided.\n\n* For paired-end read files, in order to successfully run alignment\
  \ with STAR, this tool adds the appropriate **paired-end** metadata field in the\
  \ output FASTQ files.\n\n###Common Issues and Important Notes\n\n* None\n\n###Performance\
  \ Benchmarking\n\nBelow is a table describing runtimes and task costs of **GATK\
  \ SamToFastq** for a couple of different samples, executed on the AWS cloud instances:\n\
  \n| Experiment type |  Input size | Paired-end | # of reads | Read length | Duration\
  \ |  Cost | Instance (AWS) | \n|:--------------:|:------------:|:--------:|:-------:|:---------:|:----------:|:------:|:------:|\n\
  |     RNA-Seq     |  1.9 GB |     Yes    |     16M     |     101     |   4min  \
  \ | ~0.03$ | c4.2xlarge (8 CPUs) | \n|     RNA-Seq     |  5.7 GB |     Yes    |\
  \     50M     |     101     |   7min   | ~0.04$ | c4.2xlarge (8 CPUs) | \n|    \
  \ RNA-Seq     | 9.5 GB |     Yes    |     82M    |     101     |  10min  | ~0.07$\
  \ | c4.2xlarge (8 CPUs) | \n|     RNA-Seq     | 19 GB |     Yes    |     164M  \
  \  |     101     |  20min  | ~0.13$ | c4.2xlarge (8 CPUs) |\n\n*Cost can be significantly\
  \ reduced by using **spot instances**. Visit the [Knowledge Center](https://docs.sevenbridges.com/docs/about-spot-instances)\
  \ for more details.*\n\n\n###References\n\n[1] [GATK SamToFastq](https://software.broadinstitute.org/gatk/documentation/tooldocs/4.0.12.0/picard_sam_SamToFastq)"
id: uros_sipetic/gatk-4-1-0-0-demo/gatk-samtofastq-4-1-0-0/15
inputs:
- doc: 'The action that should be taken with clipped reads: ''X'' means the reads
    and qualities should be trimmed at the clipped position; ''N'' means the bases
    should be changed to Ns in the clipped region; and any integer means that the
    base qualities should be set to that value in the clipped region.'
  id: clipping_action
  inputBinding:
    position: 5
    prefix: --CLIPPING_ACTION
    shellQuote: false
  label: Clipping action
  sbg:altPrefix: -CLIP_ACT
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'null'
  type: string?
- doc: The attribute that stores the position at which the SAM record should be clipped.
  id: clipping_attribute
  inputBinding:
    position: 5
    prefix: --CLIPPING_ATTRIBUTE
    shellQuote: false
  label: Clipping attribute
  sbg:altPrefix: -CLIP_ATTR
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'null'
  type: string?
- doc: When performing clipping with the CLIPPING_ATTRIBUTE and CLIPPING_ACTION parameters,
    ensure that the resulting reads after clipping are at least CLIPPING_MIN_LENGTH
    bases long. If the original read is shorter than CLIPPING_MIN_LENGTH then the
    original read length will be maintained.
  id: clipping_min_length
  inputBinding:
    position: 5
    prefix: --CLIPPING_MIN_LENGTH
    shellQuote: false
  label: Clipping min length
  sbg:altPrefix: -CLIP_MIN
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: '0'
  type: int?
- doc: Compress output FASTQ files per read group using gzip and append a .gz extension
    to the file names. Cannot be used in conjuction with argument(s) FASTQ (F) SECOND_END_FASTQ
    (F2) UNPAIRED_FASTQ (FU).
  id: compress_outputs_per_rg
  inputBinding:
    position: 5
    prefix: --COMPRESS_OUTPUTS_PER_RG
    shellQuote: false
  label: Compress outputs per RG
  sbg:altPrefix: -GZOPRG
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'false'
  type: boolean?
- doc: Compression level for all compressed files created (e.g. BAM and VCF).
  id: compression_level
  inputBinding:
    position: 5
    prefix: --COMPRESSION_LEVEL
    shellQuote: false
  label: Compression level
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: '2'
  type: int?
- doc: Include non-PF reads from the SAM file into the output FASTQ files. PF means
    'passes filtering'. Reads whose 'not passing quality controls' flag is set are
    non-PF reads. See GATK Dictionary for more info.
  id: include_non_pf_reads
  inputBinding:
    position: 5
    prefix: --INCLUDE_NON_PF_READS
    shellQuote: false
  label: Include non PF reads
  sbg:altPrefix: -NON_PF
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'false'
  type: boolean?
- doc: If true, include non-primary alignments in the output. Support of non-primary
    alignments in SamToFastq is not comprehensive, so there may be exceptions if this
    is set to true and there are paired reads with non-primary alignments.
  id: include_non_primary_alignments
  inputBinding:
    position: 5
    prefix: --INCLUDE_NON_PRIMARY_ALIGNMENTS
    shellQuote: false
  label: Include non primary alignments
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'false'
  type: boolean?
- doc: Input SAM/BAM file to extract reads from.
  id: in_alignments
  inputBinding:
    position: 5
    prefix: --INPUT
    shellQuote: false
  label: Input SAM/BAM file
  sbg:altPrefix: -I
  sbg:category: Required Arguments
  sbg:fileTypes: SAM, BAM
  type: File
- doc: Will generate an interleaved FASTQ if paired, each line will have /1 or /2
    to describe which end it came from.
  id: interleave
  inputBinding:
    position: 5
    prefix: --INTERLEAVE
    shellQuote: false
  label: Interleave
  sbg:altPrefix: -INTER
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'false'
  type: boolean?
- doc: This input allows a user to set the desired overhead memory when running a
    tool or adding it to a workflow. This amount will be added to the Memory per job
    in the Memory requirements section but it will not be added to the -Xmx parameter
    leaving some memory not occupied which can be used as stack memory (-Xmx parameter
    defines heap memory). This input should be defined in MB (for both the platform
    part and the -Xmx part if Java tool is wrapped).
  id: memory_overhead_per_job
  label: Memory overhead per job
  sbg:category: Platform Options
  type: int?
- doc: This input allows a user to set the desired memory requirement when running
    a tool or adding it to a workflow. This value should be propagated to the -Xmx
    parameter too.This input should be defined in MB (for both the platform part and
    the -Xmx part if Java tool is wrapped).
  id: memory_per_job
  label: Memory per job
  sbg:category: Platform Options
  sbg:toolDefaultValue: 2048 MB
  type: int?
- doc: Output a FASTQ file per read group (two FASTQ files per read group if the group
    is paired). Cannot be used in conjuction with argument(s)FASTQ (F) SECOND_END_FASTQ
    (F2) UNPAIRED_FASTQ (FU).
  id: output_per_rg
  inputBinding:
    position: 5
    prefix: --OUTPUT_PER_RG
    shellQuote: false
  label: Output per RG
  sbg:altPrefix: -OPRG
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'false'
  type: boolean?
- doc: End-trim reads using the phred/bwa quality trimming algorithm and this quality.
  id: quality
  inputBinding:
    position: 5
    prefix: --QUALITY
    shellQuote: false
  label: Quality
  sbg:altPrefix: -Q
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'null'
  type: int?
- doc: Re-reverse bases and qualities of reads with negative strand flag set before
    writing them to FASTQ.
  id: re_reverse
  inputBinding:
    position: 5
    prefix: --RE_REVERSE
    shellQuote: false
  label: Re reverse
  sbg:altPrefix: -RC
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'true'
  type:
  - 'null'
  - name: re_reverse
    symbols:
    - 'true'
    - 'false'
    type: enum
- doc: The maximum number of bases to write from read 1 after trimming. If there are
    fewer than this many bases left after trimming, all will be written. If this value
    is null then all bases left after trimming will be written.
  id: read1_max_bases_to_write
  inputBinding:
    position: 5
    prefix: --READ1_MAX_BASES_TO_WRITE
    shellQuote: false
  label: Read1 max bases to write
  sbg:altPrefix: -R1_MAX_BASES
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'null'
  type: int?
- doc: The number of bases to trim from the beginning of read 1.
  id: read1_trim
  inputBinding:
    position: 5
    prefix: --READ1_TRIM
    shellQuote: false
  label: Read1 trim
  sbg:altPrefix: -R1_TRIM
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: '0'
  type: int?
- doc: The maximum number of bases to write from read 2 after trimming. If there are
    fewer than this many bases left after trimming, all will be written. If this value
    is null then all bases left after trimming will be written.
  id: read2_max_bases_to_write
  inputBinding:
    position: 5
    prefix: --READ2_MAX_BASES_TO_WRITE
    shellQuote: false
  label: Read2 max bases to write
  sbg:altPrefix: -R2_MAX_BASES
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: 'null'
  type: int?
- doc: The number of bases to trim from the beginning of read 2.
  id: read2_trim
  inputBinding:
    position: 5
    prefix: --READ2_TRIM
    shellQuote: false
  label: Read2 trim
  sbg:altPrefix: -R2_TRIM
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: '0'
  type: int?
- doc: The read group tag (PU or ID) to be used to output a FASTQ file per read group.
  id: rg_tag
  inputBinding:
    position: 5
    prefix: --RG_TAG
    shellQuote: false
  label: RG tag
  sbg:altPrefix: -RGT
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: PU
  type: string?
- doc: Validation stringency for all SAM files read by this program. Setting stringency
    to silent can improve performance when processing a BAM file in which variable-length
    data (read, qualities, tags) do not otherwise need to be decoded.
  id: validation_stringency
  inputBinding:
    position: 5
    prefix: --VALIDATION_STRINGENCY
    shellQuote: false
  label: Validation stringency
  sbg:category: Optional Arguments
  sbg:toolDefaultValue: STRICT
  type:
  - 'null'
  - name: validation_stringency
    symbols:
    - STRICT
    - LENIENT
    - SILENT
    type: enum
- doc: Output file name prefix.
  id: output_prefix
  label: Output prefix
  sbg:category: Optional Arguments
  type: string?
- doc: Compress output file(s).
  id: compress_outputs
  label: Compress output file(s)
  sbg:category: Optional parameters
  sbg:toolDefaultValue: 'false'
  type: boolean?
- doc: CPU per job.
  id: cpu_per_job
  label: CPU per job
  sbg:category: Platform Options
  sbg:toolDefaultValue: '1'
  type: int?
label: GATK SamToFastq
outputs:
- doc: Output FASTQ file(s).
  id: out_reads
  label: Output FASTQ file(s)
  outputBinding:
    glob: "${\n    var output_ext = inputs.compress_outputs ? \".fastq.gz\" : \".fastq\"\
      ;\n    var interleave = inputs.interleave;\n    if (!inputs.outputs_by_readgroup)\n\
      \    {\n        if (interleave)\n        {\n            return \"*interleaved\"\
      \ + output_ext;\n        }\n        else\n        {\n            return [\"\
      *pe_1\" + output_ext, \"*pe_2\" + output_ext, \"*se\" + output_ext];\n     \
      \   }\n\n    }\n    else\n    {\n        return \"*\" + output_ext;\n    }\n\
      }"
    outputEval: "${ \n    self = [].concat(self)\n    \n    function getPairedEnd(filename)\n\
      \    {\n        if (filename.lastIndexOf(\".fastq\") !== 0 && filename[filename.lastIndexOf(\"\
      .fastq\") - 2 ]==\"_\") \n        {\n            return filename[filename.lastIndexOf(\"\
      .fastq\") - 1 ];\n        } \n        else \n        {\n            return \"\
      \";\n        }\n    }\n    \n    var out = inheritMetadata(self,inputs.in_alignments);\n\
      \    for (var i=0; i < out.length; i++)\n    {\n        out[i].metadata['paired_end']\
      \ = getPairedEnd(out[i].path);\n    }\n    \n    return out;\n}"
  sbg:fileTypes: FASTQ, FASTQ.GZ
  type: File[]?
- doc: Unpaired reads.
  id: unmapped_reads
  label: Unpaired reads
  outputBinding:
    glob: "${\n    var output_ext = inputs.compress_outputs ? \".fastq.gz\" : \".fastq\"\
      ;\n    var interleave = inputs.interleave;\n    if (!inputs.outputs_by_readgroup)\n\
      \    {\n        if (!interleave)\n        {\n            return \"*unpaired\"\
      \ + output_ext;\n        }\n        else \n        {\n             return \"\
      \"; \n        }      \n    }\n  else {\n       return \"\";  \n  \n  }\n}"
    outputEval: $(inheritMetadata(self, inputs.in_alignments))
  sbg:fileTypes: FASTQ, FASTQ.GZ
  type: File?
requirements:
- class: ShellCommandRequirement
- class: ResourceRequirement
  coresMin: "${\n    return inputs.cpu_per_job ? inputs.cpu_per_job : 1;\n}"
  ramMin: "${\n    var memory = 4096;\n    if (inputs.memory_per_job) \n    {\n  \
    \      memory = inputs.memory_per_job;\n    }\n    if (inputs.memory_overhead_per_job)\n\
    \    {\n        memory += inputs.memory_overhead_per_job;\n    }\n    return memory;\n\
    }"
- class: DockerRequirement
  dockerPull: images.sbgenomics.com/stefan_stojanovic/gatk:4.1.0.0
- class: InitialWorkDirRequirement
  listing: []
- class: InlineJavascriptRequirement
  expressionLib:
  - "var updateMetadata = function(file, key, value) {\n    file['metadata'][key]\
    \ = value;\n    return file;\n};\n\n\nvar setMetadata = function(file, metadata)\
    \ {\n    if (!('metadata' in file))\n        file['metadata'] = metadata;\n  \
    \  else {\n        for (var key in metadata) {\n            file['metadata'][key]\
    \ = metadata[key];\n        }\n    }\n    return file\n};\n\nvar inheritMetadata\
    \ = function(o1, o2) {\n    var commonMetadata = {};\n    if (!Array.isArray(o2))\
    \ {\n        o2 = [o2]\n    }\n    for (var i = 0; i < o2.length; i++) {\n   \
    \     var example = o2[i]['metadata'];\n        for (var key in example) {\n \
    \           if (i == 0)\n                commonMetadata[key] = example[key];\n\
    \            else {\n                if (!(commonMetadata[key] == example[key]))\
    \ {\n                    delete commonMetadata[key]\n                }\n     \
    \       }\n        }\n    }\n    if (!Array.isArray(o1)) {\n        o1 = setMetadata(o1,\
    \ commonMetadata)\n    } else {\n        for (var i = 0; i < o1.length; i++) {\n\
    \            o1[i] = setMetadata(o1[i], commonMetadata)\n        }\n    }\n  \
    \  return o1;\n};\n\nvar toArray = function(file) {\n    return [].concat(file);\n\
    };\n\nvar groupBy = function(files, key) {\n    var groupedFiles = [];\n    var\
    \ tempDict = {};\n    for (var i = 0; i < files.length; i++) {\n        var value\
    \ = files[i]['metadata'][key];\n        if (value in tempDict)\n            tempDict[value].push(files[i]);\n\
    \        else tempDict[value] = [files[i]];\n    }\n    for (var key in tempDict)\
    \ {\n        groupedFiles.push(tempDict[key]);\n    }\n    return groupedFiles;\n\
    };\n\nvar orderBy = function(files, key, order) {\n    var compareFunction = function(a,\
    \ b) {\n        if (a['metadata'][key].constructor === Number) {\n           \
    \ return a['metadata'][key] - b['metadata'][key];\n        } else {\n        \
    \    var nameA = a['metadata'][key].toUpperCase();\n            var nameB = b['metadata'][key].toUpperCase();\n\
    \            if (nameA < nameB) {\n                return -1;\n            }\n\
    \            if (nameA > nameB) {\n                return 1;\n            }\n\
    \            return 0;\n        }\n    };\n\n    files = files.sort(compareFunction);\n\
    \    if (order == undefined || order == \"asc\")\n        return files;\n    else\n\
    \        return files.reverse();\n};"
  - "\nvar setMetadata = function(file, metadata) {\n    if (!('metadata' in file))\
    \ {\n        file['metadata'] = {}\n    }\n    for (var key in metadata) {\n \
    \       file['metadata'][key] = metadata[key];\n    }\n    return file\n};\n\n\
    var inheritMetadata = function(o1, o2) {\n    var commonMetadata = {};\n    if\
    \ (!Array.isArray(o2)) {\n        o2 = [o2]\n    }\n    for (var i = 0; i < o2.length;\
    \ i++) {\n        var example = o2[i]['metadata'];\n        for (var key in example)\
    \ {\n            if (i == 0)\n                commonMetadata[key] = example[key];\n\
    \            else {\n                if (!(commonMetadata[key] == example[key]))\
    \ {\n                    delete commonMetadata[key]\n                }\n     \
    \       }\n        }\n    }\n    if (!Array.isArray(o1)) {\n        o1 = setMetadata(o1,\
    \ commonMetadata)\n    } else {\n        for (var i = 0; i < o1.length; i++) {\n\
    \            o1[i] = setMetadata(o1[i], commonMetadata)\n        }\n    }\n  \
    \  return o1;\n};"
sbg:appVersion:
- v1.0
sbg:categories:
- Utilities
- BAM Processing
sbg:content_hash: a21e1194e724a1f17bceabd4d2040324713c2a5c63896adcebbc777578b2bfef5
sbg:contributors:
- uros_sipetic
- nens
- veliborka_josipovic
sbg:createdBy: uros_sipetic
sbg:createdOn: 1552663400
sbg:id: uros_sipetic/gatk-4-1-0-0-demo/gatk-samtofastq-4-1-0-0/15
sbg:image_url: null
sbg:latestRevision: 15
sbg:license: Open source BSD (3-clause) license
sbg:links:
- id: https://software.broadinstitute.org/gatk/
  label: Homepage
- id: https://github.com/broadinstitute/gatk/
  label: Source Code
- id: https://github.com/broadinstitute/gatk/releases/download/4.1.0.0/gatk-4.1.0.0.zip
  label: Download
- id: https://www.ncbi.nlm.nih.gov/pubmed?term=20644199
  label: Publications
- id: https://software.broadinstitute.org/gatk/documentation/tooldocs/4.1.0.0/picard_sam_SamToFastq.php
  label: Documentation
sbg:modifiedBy: veliborka_josipovic
sbg:modifiedOn: 1561548030
sbg:project: uros_sipetic/gatk-4-1-0-0-demo
sbg:projectName: GATK 4.1.0.0 - Demo
sbg:publisher: sbg
sbg:revision: 15
sbg:revisionNotes: Added glob for single end output fastq
sbg:revisionsInfo:
- sbg:modifiedBy: uros_sipetic
  sbg:modifiedOn: 1552663400
  sbg:revision: 0
  sbg:revisionNotes: Copy of veliborka_josipovic/gatk-4-1-0-0-toolkit-dev/gatk-samtofastq-4-1-0-0/19
- sbg:modifiedBy: uros_sipetic
  sbg:modifiedOn: 1552663734
  sbg:revision: 1
  sbg:revisionNotes: Copy of veliborka_josipovic/gatk-4-1-0-0-toolkit-dev/gatk-samtofastq-4-1-0-0/20
- sbg:modifiedBy: veliborka_josipovic
  sbg:modifiedOn: 1554492676
  sbg:revision: 2
  sbg:revisionNotes: Copy of veliborka_josipovic/gatk-4-1-0-0-toolkit-dev/gatk-samtofastq-4-1-0-0/27
- sbg:modifiedBy: veliborka_josipovic
  sbg:modifiedOn: 1554493243
  sbg:revision: 3
  sbg:revisionNotes: Copy of veliborka_josipovic/gatk-4-1-0-0-toolkit-dev/gatk-samtofastq-4-1-0-0/28
- sbg:modifiedBy: veliborka_josipovic
  sbg:modifiedOn: 1554720826
  sbg:revision: 4
  sbg:revisionNotes: Copy of veliborka_josipovic/gatk-4-1-0-0-toolkit-dev/gatk-samtofastq-4-1-0-0/29
- sbg:modifiedBy: veliborka_josipovic
  sbg:modifiedOn: 1554999298
  sbg:revision: 5
  sbg:revisionNotes: Copy of veliborka_josipovic/gatk-4-1-0-0-toolkit-dev/gatk-samtofastq-4-1-0-0/30
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1557484228
  sbg:revision: 6
  sbg:revisionNotes: Updated Description
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1557745933
  sbg:revision: 7
  sbg:revisionNotes: ''
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1557918579
  sbg:revision: 8
  sbg:revisionNotes: 'v32: [input]'
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1557919927
  sbg:revision: 9
  sbg:revisionNotes: v5->update
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1558002424
  sbg:revision: 10
  sbg:revisionNotes: output required
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1558015833
  sbg:revision: 11
  sbg:revisionNotes: unmapped_reads required
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1558022954
  sbg:revision: 12
  sbg:revisionNotes: strict js for glob
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1558023482
  sbg:revision: 13
  sbg:revisionNotes: strict javascript for unmapped_reads
- sbg:modifiedBy: nens
  sbg:modifiedOn: 1558354170
  sbg:revision: 14
  sbg:revisionNotes: return '';
- sbg:modifiedBy: veliborka_josipovic
  sbg:modifiedOn: 1561548030
  sbg:revision: 15
  sbg:revisionNotes: Added glob for single end output fastq
sbg:sbgMaintained: false
sbg:toolAuthor: Broad Institute
sbg:toolkit: GATK
sbg:toolkitVersion: 4.1.0.0
sbg:validationErrors: []
